### ISSUE TITLE
Add new variable interactive_pipestatus

### ISSUE THREAD

**Issue by IlanCosman** [fish-shell/fish-shell#7830]
Currently there's a ton of hidden state sorrounding the `status` and `pipestatus` variables. For example, `fish_prompt`, `fish_right_prompt`, and `fish_mode_prompt` all use the status that was last set from the commandline. I propose that we reduce this hidden state, while at the same time allowing other functions to have this ability. 

As an example use case, I want to create a `fish_bind_mode` variable handler that reacts to the previous commandline status. 

```fish
function on_fish_bind_mode --on-variable fish_bind_mode
    echo $pipestatus
end
```

Unfortunately the status from the bind functions carries over instead, so all we get are `0` and `1`.

Adding an `interactive_pipestatus` variable would also let us reduce the unergonomic thing that every `fish_prompt` in the world does at the start, namely saving the commandline status/pipestatus:

```fish
set -l last_status $pipestatus
```

**Comment by IlanCosman** [comment]
Also, this is probably not too hard so I'm happy to attempt to implement this myself if the idea is accepted.

**Comment by ridiculousfish** [comment]
I think this makes a lot of sense. Wouldn't we want `$interactive_status` in addition to `$interactive_pipestatus`? (The main issue there is `not` which applies to a whole pipeline - `not true | false` gets a status of 0 ('not false') but pipestatus ignores the not)

**Comment by IlanCosman** [comment]
Oh, yeah, I forgot about `not`. Yes, we would also want an `$interactive_status` variable. 😄

**Comment by krobelus** [comment]
Yeah it sounds good to make this more flexible.
I think changing `$status` in `fish_prompt` would be a pretty harsh compatibilty break.
Maybe it make sense to expose something new like extra subcommands to `builtin status` and maintain the do-what-I-mean behavior of `$status`? Something like `status {,interactive-}{status,pipestatus}`.

**Comment by IlanCosman** [comment]
> I think changing `$status` in `fish_prompt` would be a pretty harsh compatibility break.

Agreed, but I'd say deprecate it and remove it a few versions down the line. The vast majority of people use either the default prompts under our control, or the handful of popular third-party prompts like Tide, Starship, Pure, Bobthefish. Those are under enough active development that it wouldn't be an issue.

> Maybe it make sense to expose something new like extra subcommands to builtin status

I don't think so. That would be reducing orthogonality, not increasing it. We already use variables in this space, and we don't plan to remove them, so I think adding new variables would be the best choice.

**Comment by krobelus** [comment]
> We already use variables in this space

Makes sense - I think it's best to add `$interactive_{status,pipestatus}` but still keep the behavior of `$status`. As far as I can tell it only has the special behavior when we execute a prompt - I don't think that hurts anyone?

**Comment by IlanCosman** [comment]
Alright, it sounds like this idea is accepted. Any ideas as to where I'd start? I did a little poking around and this seems much more difficult than I had originally thought 😄

**Comment by krobelus** [comment]
Take a look at how `$status_generation` is implemented. Maybe you can mimic that?
Add a new field in `library_data_t` and update it whenever we update `status_count`.. I didn't test it though.

**Comment by IlanCosman** [comment]
Hi, sorry that I haven't gotten around to this, I still intend to do so at some point 😞 Since the idea has been accepted, can we mark this as `fish-future`, just for safekeeping?

**Comment by ilyagr** [comment]
It seems to me that the following might just work already (in master, at least):

```fish
function __fish_set_interactive_pipestatus  --on-event fish_postexec
       set -g interactive_pipestatus $pipestatus
       # TODO: Set `interactive_status` from `interactive_pipestatus`
end
```

Perhaps we should merely make this hook a part of fish's default init sequence and document it? Or is there some known problem? I expected some, but they didn't materialize. In particular, I was pleasantly surprised that this did *not* mess up the usual `$pipestatus`, and prompts relying on that still seemed to work.

Also, it seems like a good workaround for others even if it's not implemented in `fish`.

**Update:** This seems to work for me:

```console
$ set -g ilya_pipestatus 0
$ function __ilya_set_pipestatus  --on-event fish_postexec
      set -g ilya_pipestatus $pipestatus
  end

$ function fish_prompt
      echo '>' pipestatus: $pipestatus
      echo '>' ilya_pipestatus: $ilya_pipestatus
      echo '$ '
  end
> pipestatus: 0
> ilya_pipestatus: 0
$ false | true
> pipestatus: 1 0
> ilya_pipestatus: 1 0
$ 
```

<details><summary>Obsolete text</summary>

(**Update:** I think I figured this out, will probably turn it into a PR for fish-async prompt once I decide on how to best do that) I'm having some trouble making it work with `fish-async-prompt`, but I'll try to fix that. With fish-async-prompt, I seem to get:

```
> pipestatus: 1 0
> ilya_pipestatus: 0
```

for some reason.

----------

</details>

**Update:** This doesn't quite work with `fish-async-prompt`, but a similar trick from https://github.com/acomagu/fish-async-prompt/pull/90 seems to (regardless of whether this is added to fish proper)

**Comment by josephlou5** [comment]
> It seems to me that the following might just work already (in master, at least):

I'm on fish v4.5.0 and @ilyagr's approach worked great! Not sure if his "Update: This doesn't quite work with `fish-async-prompt`" is still relevant, as that worked for me as well (make sure to add the variable to `async_prompt_inherit_variables` if needed).

> In particular, I was pleasantly surprised that this did _not_ mess up the usual `$pipestatus`, and prompts relying on that still seemed to work.

This may be because `set` in assignment mode does not alter `$status` ([docs](https://fishshell.com/docs/current/cmds/set.html#exit-status)), so presumably it does not alter `$pipestatus` either, though this isn't explicitly documented.
